ixfr.h

```diff
@@ -0,0 +1,270 @@
+/*
+ * ixfr.h -- generating IXFR responses.
+ *
+ * Copyright (c) 2021, NLnet Labs. All rights reserved.
+ *
+ * See LICENSE for the license.
+ *
+ */
+
+#ifndef _IXFR_H_
+#define _IXFR_H_
+struct nsd;
+#include "query.h"
+#include "rbtree.h"
+struct ixfr_data;
+struct zone;
+struct buffer;
+struct region;
+
+/* default for number of ixfr versions in files */
+#define IXFR_NUMBER_DEFAULT 5 /* number of versions */
+/* default for IXFR storage */
+#define IXFR_SIZE_DEFAULT 1048576 /* in bytes, 1M */
+
+/* data structure that stores IXFR contents for a zone. */
+struct zone_ixfr {
+	/* Items are of type ixfr_data. The key is old_serial.
+	 * So it can be looked up on an incoming IXFR. They are sorted
+	 * by old_serial, so the looked up and next are the versions needed.
+	 * Tree of ixfr data for versions */
+	struct rbtree* data;
+	/* total size stored at this time, in bytes,
+	 * sum of sizes of the ixfr data elements */
+	size_t total_size;
+	/* the oldest serial number in the tree, searchable by old_serial */
+	uint32_t oldest_serial;
+	/* the newest serial number in the tree, that is searchable in the
+	 * tree, so it is the old_serial of the newest data entry, that
+	 * has an even newer new_serial of that entry */
+	uint32_t newest_serial;
+	/* the newserial number of the newest ixfr data element */
+	uint32_t newest_newserial;
```

Comment:
I added it because it allows comparison with the current serial in the nsd domain tree. But that comparison does not happen from that contents. I guess I should remove it, if it is not used.